pkg/controller/migplan/registry_health_failed.go

```diff
@@ -0,0 +1,129 @@
+	"time"
+)
+
+// A registry health checker which routinely checks whether registries are healthy or not for migplan's failed migrations
+type registryHealthFailed struct {
+	hostClient client.Client
+	Interval   time.Duration
+	handler    handler.EventHandler
+	queue      workqueue.RateLimitingInterface
+	predicates []predicate.Predicate
+	podLabels  map[string]string
+}
+
+// Start the health checks
+func (r *registryHealthFailed) Start(
+	handler handler.EventHandler,
+	queue workqueue.RateLimitingInterface,
+	predicates ...predicate.Predicate) error {
+
+	r.handler = handler
+	r.queue = queue
+	r.predicates = predicates
+	go r.run()
+
+	return nil
+}
+
+// Enqueue a reconcile request event for migplan
+func (r *registryHealthFailed) enqueue(plan v1alpha1.MigPlan) {
+	event := event.GenericEvent{
+		Meta:   &plan.ObjectMeta,
+		Object: &plan,
+	}
+	for _, predicate := range r.predicates {
+		if !predicate.Generic(event) {
+			return
+		}
+	}
+
+	r.handler.Generic(event, r.queue)
+}
+
+//Run the health checks for registry pods
+func (r *registryHealthFailed) run() {
+	//List all the migplans that have migrations in failed state using the hostClient
+	//Now using srcClient and destClient for each migplan find the registry pods, if registry container is not ready then enqueue this migplan
+	//repeat all the above steps for all the plans for both clusters
+
+	for {
+		time.Sleep(r.Interval)
+		planList, err := v1alpha1.ListFailedPlans(r.hostClient)
+		if err != nil {
+			log.Trace(err)
+			return
+		}
+		if planList != nil {
+			for _, plan := range planList {
+				//TODO avoid race condition check
+				srcCluster, err := plan.GetSourceCluster(r.hostClient)
+				if err != nil {
+					log.Trace(err)
+					//TODO Error condition would result in loosing the watch util the pod restarts
+					return
+				}
+
+				if !srcCluster.Status.IsReady() {
+					log.Info("Cannot check registry pod health, cluster is not ready", srcCluster.Name, plan.Name)
+					return
+				}
+
+				srcClient, err := srcCluster.GetClient(r.hostClient)
+				if err != nil {
+					return
+				}
+
+				destCluster, err := plan.GetDestinationCluster(r.hostClient)
+				if err != nil {
+					log.Trace(err)
+					return
+				}
+
+				if !destCluster.Status.IsReady() {
+					log.Info("Cannot check registry pod health, cluster is not ready", destCluster.Name, plan.Name)
+					return
+				}
+
+				destClient, err := destCluster.GetClient(r.hostClient)
+				if err != nil {
+					return
+				}
+
+				isSrcRegistryPodUnhealthy, _, err := isRegistryPodUnHealthy(&plan, srcClient)
+				if err != nil {
+					log.Trace(err)
+					return
+				}
+
+				isDestRegistryPodUnhealthy, _, err := isRegistryPodUnHealthy(&plan, destClient)
+				if err != nil {
+					log.Trace(err)
+					return
+				}
+
+				switch {
+				//enqueue a reconcile event when the src registry pod is healthy and the plan is not ready
+				case !isSrcRegistryPodUnhealthy && !plan.Status.HasCondition(RegistriesHealthy):
```

Comment:
you would need this `&& !plan.Status.HadCondition(Ready)` to handle operator opdates right?

Reply:
Added